### DISCUSSION TITLE
how to get render result with a texture or a pixelbuffer

### DISCUSSION THREAD

**Discussion by Tony-qty** [google/filament#4854] [Q&A] [answered]
i want to get  the render result  by this way. but it seems poor performance. 

 if (m_renderer->beginFrame(m_swap_chain)) {
        m_renderer->render(m_view);
        PixelBufferDescriptor pbd(buffer, buf_size, PixelDataFormat::RGBA, PixelDataType::UBYTE,
                                  sPixelBufferDescriptorCallback, nullptr);
        m_renderer->readPixels(0, 0, width, height, std::move(pbd));
        m_renderer->endFrame();
        m_engine->flushAndWait();
    } 

who can gives me more idea

**Comment by romainguy** [comment, marked as answer]
You can use a RenderTarget to render directly into a texture.

**Reply by Tony-qty** [reply]
thanks for your idea；
but，my texture is a OpenGL texture, which i use to mix with other image,  not filament texture class

**Reply by romainguy** [reply]
On Android you can use a SurfaceTexture for this then.

**Reply by Tony-qty** [reply]
0.0 emmm，what can I do on ios.

**Reply by romainguy** [reply]
@bejado would know, I’m sure there’s an equivalent.

**Reply by Tony-qty** [reply]
@bejado bro, can you hear me?

**Reply by bejado** [reply]
Your best bet is probably using Filament to render into a `CVPixelBuffer`. You can do this by enabling the [`CONFIG_APPLE_CVPIXELBUFFER` flag](https://github.com/google/filament/blob/41a809368b37f33e562d767e1aca2a0c3a166749/filament/include/filament/SwapChain.h#L178) on a `SwapChain` and passing in a `CVPixelBufferRef` as the `nativeWindow` argument to `createSwapChain`.

If you also need to render into the buffer from OpenGL, you should be able to create an associated OpenGL texture via `CVOpenGLESTextureCacheCreateTextureFromImage`.

**Reply by Tony-qty** [reply]
thanks for your advice, that's very userful

**Reply by Tony-qty** [reply]
> Your best bet is probably using Filament to render into a `CVPixelBuffer`. You can do this by enabling the [`CONFIG_APPLE_CVPIXELBUFFER` flag](https://github.com/google/filament/blob/41a809368b37f33e562d767e1aca2a0c3a166749/filament/include/filament/SwapChain.h#L178) on a `SwapChain` and passing in a `CVPixelBufferRef` as the `nativeWindow` argument to `createSwapChain`.
> 
> If you also need to render into the buffer from OpenGL, you should be able to create an associated OpenGL texture via `CVOpenGLESTextureCacheCreateTextureFromImage`.

there is something wrong when I use CVPixelBuffer as the `nativeWindow` argument to `createSwapChain`.
![image](https://user-images.githubusercontent.com/7675144/142363654-f0c3cde0-cae7-46f8-bcb7-3cf46896e1dc.png)

**Reply by bejado** [reply]
You'll need to use the Metal backend to render into a `CVPixelBuffer`, it's not supported with OpenGL.

**Reply by Tony-qty** [reply]
> You'll need to use the Metal backend to render into a `CVPixelBuffer`, it's not supported with OpenGL.
If I use OpenGL as backend, how to run fliament without main thread

**Reply by bejado** [reply]
You don't need to run Filament on the main thread, but you must be consistent. All Filament APIs must be called from the same thread.

**Reply by dandingol03** [reply]
@Tony-qty Hi, can you share the code? Recently i was trying to generate a video which can record the filament stuff. My idea is generate the `CVPixelBuffer` and render things to it, and write to a file.

**Reply by Tony-qty** [reply]
Here is the codes:

first you must generate a  pixelBuffer:

- (CVPixelBufferRef)generatePixelbuffer {
    
    NSDictionary *pixelAttributes = @{(id)kCVPixelBufferIOSurfacePropertiesKey : @{}};
    CVPixelBufferRef newPixelBuffer = NULL;
    CVReturn result = CVPixelBufferCreate(kCFAllocatorDefault,
                                          self.config.width,
                                          self.config.height,
                                          kCVPixelFormatType_32BGRA,
                                          (__bridge CFDictionaryRef)(pixelAttributes),
                                          &newPixelBuffer);
    
    return newPixelBuffer;
}

then, setup your engine with pixelbuffer

    pb = [self generatePixelbuffer];
    renderProcessor->Setup(self.config.width, self.config.height, pb);

meanwhile ，you need change your swapchain to "SWAP_CHAIN_CONFIG_APPLE_CVPIXELBUFFER"

    m_swap_chain = m_engine->createSwapChain(window, SWAP_CHAIN_CONFIG_APPLE_CVPIXELBUFFER);

at last , when your filament EndFrame , you can read pixel from the pixelbuffer

**Reply by dandingol03** [reply]
thanks a lot~

**Reply by simonsland** [reply]
// 1. create pixelBuffer
    NSDictionary *pixelAttributes = @{(id)kCVPixelBufferIOSurfacePropertiesKey : @{}};
    CVPixelBufferRef newPixelBuffer = NULL;
    int width = self.bounds.size.width;
    int height = self.bounds.size.height;
    CVReturn result = CVPixelBufferCreate(kCFAllocatorDefault, width, height, kCVPixelFormatType_32BGRA, (__bridge CFDictionaryRef)(pixelAttributes), &newPixelBuffer);
    
    // 2. createSwapChain
    _swapChain = _engine->createSwapChain(newPixelBuffer, SwapChain::CONFIG_APPLE_CVPIXELBUFFER | SwapChain::CONFIG_READABLE);
    
    if (_renderer->beginFrame(_swapChain)) {
        _renderer->render(_view);
        _renderer->endFrame();
        
        // 3. readPixels
        CVPixelBufferRef pixelBuffer = newPixelBuffer;
        CIImage *ciImage = [CIImage imageWithCVPixelBuffer:pixelBuffer];
        CIContext *temporaryContext = [CIContext contextWithOptions:nil];
        CGImageRef imgRef = [temporaryContext createCGImage:ciImage fromRect:CGRectMake(0, 0,CVPixelBufferGetWidth(pixelBuffer),CVPixelBufferGetHeight(pixelBuffer))];
        UIImage *uiimg = [UIImage imageWithCGImage:imgRef];
        CGImageRelease(imgRef);
    }
    
    -------------
    
Hi, recently, I've been tried also. As shown above, but it doesn't work, the PixelBuffer seems unchanged. It's there anything wrong with the codes? @Tony-qty

**Reply by bejado** [reply]
@starccode You may need to insert a delay after calling `endFrame` to ensure the render has had time to complete. You can also use the [`setFrameCompleted` API](https://github.com/google/filament/blob/bb460d78d8a9168434cc1c333ea9505ad25a4a42/filament/include/filament/SwapChain.h#L225) to be notified when a frame has finished.

**Reply by simonsland** [reply]
I've tried and it works, thanks a lot!!

**Reply by dandingol03** [reply]
works fine!

**Reply by dandingol03** [reply]
works fine when backend is `metal`, crash  on `opengl` backend

**Reply by dandingol03** [reply]
@Tony-qty hi, which backend you choose. This code crash when i use `opengl` to create engine.
